pkg/expressions/function/functions.go

```diff
@@ -498,12 +559,79 @@ func getCommit(
 	}
 }
 
-// getImage returns a function that finds container images based on repository
+// getCommitFromDiscoveredArtifacts returns a function that finds Git commits based on repository URL.
+func getCommitFromDiscoveredArtifacts(
+	ctx context.Context,
+	wh *kargoapi.Warehouse,
+	artifacts *kargoapi.DiscoveredArtifacts,
+) exprFn {
+	return func(a ...any) (any, error) {
+		if len(a) != 1 {
+			return nil, fmt.Errorf("expected 1 argument, got %d", len(a))
+		}
+
+		repoURL, ok := a[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("first argument must be string, got %T", a[0])
+		}
+
+		repoURL = urls.NormalizeGit(repoURL)
+
+		logger := logging.LoggerFromContext(ctx).WithValues(
+			"repoURL", repoURL,
+			"warehouse", wh.Name,
+		)
+
+		var latestCommit *kargoapi.DiscoveredCommit
+		for _, s := range wh.Spec.Subscriptions {
+			if s.Git != nil && urls.NormalizeGit(s.Git.RepoURL) == repoURL && len(artifacts.Git) != 0 {
+				logger.Debug("number of discovered git artifacts",
+					"count", len(artifacts.Git),
+				)
+				for i, ca := range artifacts.Git {
+					if urls.NormalizeGit(ca.RepoURL) != repoURL {
+						continue
+					}
+					logger.Debug("checking discovered git artifact",
+						"index", i,
+						"numCommits", len(ca.Commits),
+					)
+					slices.SortFunc(ca.Commits, func(a, b kargoapi.DiscoveredCommit) int {
+						return a.CreatorDate.Compare(b.CreatorDate.Time)
+					})
```

Comment:
Not sure why we're sorting. Discovery should have sorted these already. iirc the 0 element is the best match for the selection criteria.